### ISSUE TITLE
bzip2 files created by pbzip2 are not supported

### ISSUE THREAD

**Issue by alpapan** [galaxyproject/galaxy#4538]
This confused me until I realised what was happening:

For a bz2 compressed fastq file:
if upload type is set as sangerfastq (or fastq) or auto-detect then the file is uploaded succesfully and uncompressed but only a few thousand reads (800kb). 

If set as sangerfastq.bz2 it works ok but you have to manually then uncompress it.

Further if using FTP, the original file is overwritten and forever lost (see issue #4527)

I don't know if any of the above also fail with direct upload
thanks
a

**Comment by mvdbeek** [comment]
This sounds like the correct behavior, see https://galaxyproject.org/tutorials/upload/#to-gzip-or-not-to-gzip. If you assign a datatype during upload, you do exactly that. This setting is not for converting datatypes, so if you set the fastqsanger datatype on a fastqsanger.gz file your upload will be unusable.

You should not ever need to decompress compressed fastq files. Most tools allow compressed fastq input (though gz support is more common than bz2), and if they don't galaxy will automatically decompress files.

Files being moved away from the FTP folder is also normal, and galaxy administrators can control whether that should happen or not with the `ftp_upload_purge` setting (https://github.com/galaxyproject/galaxy/blob/dev/config/galaxy.ini.sample#L1192).

**Comment by alpapan** [comment]
Hello

Sorry, let me rephrase
When I upload a FASTQ.BZ2 using autodetect or setting it as fastq, galaxy uploads and and automatically **part of it** (small file size), keeping only about "4,777 sequences". No error is reported.

![image](https://user-images.githubusercontent.com/608484/30044233-5fc11974-923f-11e7-9cfe-03af32d2bdd5.png)

If i select datatype fastq.bz2 then everything is ok 
![image](https://user-images.githubusercontent.com/608484/30044284-9077ba46-923f-11e7-8b09-2b0373e78b7f.png)

but you have to convert fastq.bz2 -> fastq. If you do that, filesize is correct and the right number of sequences are there.
![image](https://user-images.githubusercontent.com/608484/30044412-44821d56-9240-11e7-9151-4a46ae6ea036.png)


FYI, here is the upload window:
![image](https://user-images.githubusercontent.com/608484/30044265-81d73dc2-923f-11e7-898c-31579080946c.png)

i wonder if it is an issue here because CHUNK_SIZE of 1 mb sounds about 4k sequences.......

CHUNK_SIZE = 2 ** 20  # 1Mb
                        fd, uncompressed = tempfile.mkstemp( prefix='data_id_%s_upload_bunzip2_' % dataset.dataset_id, dir=os.path.dirname( output_path ), text=False )
                        bzipped_file = bz2.BZ2File( dataset.path, 'rb' )
                        while 1:
                            try:
                                chunk = bzipped_file.read( CHUNK_SIZE )  


thanks
a

**Comment by mvdbeek** [comment]
This is all behaving as intended. You do not have to explicitly decompress datasets, Galaxy will do this for you if a downstream tool does not support compressed datatypes. Galaxy will not prevent you from assigning the wrong datatype (as you do when you say that your compressed file is a fastqsanger file).

**Comment by nsoranzo** [comment]
But @alpapan said that the problem happens also when leaving it to autodetect.

**Comment by mvdbeek** [comment]
Oh right, that shouldn't happen. I just tested this on usegalaxy.org (running 17.05) and it works fine.
@alpapan, could you double-check that you are not getting the complete, decompressed version of the fastq file if you upload with auto settings. If you get a truncated version of this file it would be great if you can make this file available for us to test.

**Comment by alpapan** [comment]
Hello

I've tried with any fastq file (17.05) e.g. 
https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR4413681

And indeed it is truncated. I first noticed a few days ago because my alignments took seconds instead of hours....

A) I can also confirm i just made a fresh new installation of galaxy (17.05) and it is still happening. 
B) It was the ftp upload approach
C) But most importantly, it was done with an SGE gridengine but the upload1 tool is running locally

        <tool id="upload1" destination="local"/>
        <tool id="CONVERTER_fastqbz2_to_fastq" destination="SGE"/>

i wonder if any of these may be the issue...
a

**Comment by mvdbeek** [comment]
Probably, yes. Can you verify direct uploads work ?

**Comment by alpapan** [comment]
Hello

Using autodetect:
1) i tried direct upload with a 1.5gb bz2 compressed file (with no other changes to SGE) and didn't work:  dataset_65.dat 879K
2) I removed the influence of SGE
```
        <tool id="CONVERTER_fastqbz2_to_fastq" destination="local"/>
```
and tried to re-upload a local file and same problem
3) there is no issue if it is not compressed

**Comment by mvdbeek** [comment]
> I removed the influence of SGE
>         <tool id="CONVERTER_fastqbz2_to_fastq" destination="local"/>

This converter is not used during upload.

**Comment by alpapan** [comment]
then why is the bzip running on the cluster while the upload is local?

there are no other tools with destination sge

**Comment by alpapan** [comment]
BTW, with FTP it also overwrites the original file on the FTP (to 800kb) which is kinda terrible

**Comment by alpapan** [comment]
i was wrong the bzip doesnt run at all. i confused it to when i run the CONVERTER

so the issue with upload1...

**Comment by mvdbeek** [comment]
> BTW, with FTP it also overwrites the original file on the FTP (to 800kb) which is kinda terrible

Yes, that will be fixed with https://github.com/galaxyproject/galaxy/pull/4539

**Comment by alpapan** [comment]
Ok, so i tried with gzip. Unlike bz2 files which were processed in < 1 minute, the gz ones take much longer to process (hasn't finished yet it will take a dozen minutes or so)

I wonder if the issue may be that my bz2 files are compressed with pbzip2. Could someone check please? time for me to leave the office (Sydney time)...

a

**Comment by alpapan** [comment]
gzip worked fine

**Comment by mvdbeek** [comment]
> I wonder if the issue may be that my bz2 files are compressed with pbzip2.

Yep, that seems to be the problem. I can confirm that something is going wrong for files compressed by pbzip2 and https://docs.python.org/2/library/bz2.html#de-compression-of-files states that the bz2 module does not support pbzip2 files. They suggest using https://pypi.python.org/pypi/bz2file.

**Comment by mvdbeek** [comment]
Thanks @alpapan for insisting, I shouldn't have jumped to conclusions!
